cpp/dolfin/mesh/Mesh.cpp

```diff
@@ -48,79 +48,157 @@ Eigen::ArrayXd cell_r(const mesh::Mesh& mesh)
+                                  Eigen::RowMajor>>
+        cell_nodes)
 {
+  // Get set of all global points needed locally
   const std::int32_t num_cells = cell_nodes.rows();
   const std::int32_t num_nodes_per_cell = cell_nodes.cols();
+  std::set<std::int64_t> gi_set;
+  for (std::int32_t c = 0; c < num_cells; ++c)
+    for (std::int32_t v = 0; v < num_nodes_per_cell; ++v)
+      gi_set.insert(cell_nodes(c, v));
 
-  // Cell points in local indexing
-  Eigen::Array<std::int32_t, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
-      cell_nodes_local(num_cells, num_nodes_per_cell);
-
-  // Loop over cells to build local-to-global map for (i) vertex nodes,
-  // and (ii) then other nodes
+  return std::vector<std::int64_t>(gi_set.begin(), gi_set.end());
+}
+//-----------------------------------------------------------------------------
+std::pair<std::vector<std::int64_t>, std::array<int, 4>>
+compute_local_to_global_point_map(
+    MPI_Comm mpi_comm, int num_vertices_per_cell,
+    const std::map<std::int64_t, std::set<int>>& point_to_procs,
+    Eigen::Ref<const Eigen::Array<std::int64_t, Eigen::Dynamic, Eigen::Dynamic,
+                                  Eigen::RowMajor>>
+        cell_nodes)
+{
+  int mpi_rank = dolfin::MPI::rank(mpi_comm);
   std::vector<std::int64_t> local_to_global;
-  std::map<std::int64_t, std::int32_t> global_to_local;
-  std::int32_t num_vertices_local = 0;
-  int v0(0), v1(num_vertices_per_cell);
-  for (std::int32_t pass = 0; pass < 2; ++pass)
+  std::array<int, 4> num_vertices_local;
+
+  // Classify all nodes
+  std::set<std::int64_t> local_vertices;
+  std::set<std::int64_t> shared_vertices;
+  std::set<std::int64_t> ghost_vertices;
+  std::set<std::int64_t> non_vertex_nodes;
+
+  const std::int32_t num_cells = cell_nodes.rows();
+  const std::int32_t num_nodes_per_cell = cell_nodes.cols();
+  for (std::int32_t c = 0; c < num_cells; ++c)
   {
-    for (std::int32_t c = 0; c < num_cells; ++c)
+    // Loop over vertex nodes
+    for (std::int32_t v = 0; v < num_vertices_per_cell; ++v)
     {
-      // Loop over cell nodes
-      for (std::int32_t v = v0; v < v1; ++v)
+      // Get global node index
+      std::int64_t q = cell_nodes(c, v);
+      auto shared_it = point_to_procs.find(q);
+      if (shared_it == point_to_procs.end())
+        local_vertices.insert(q);
+      else
       {
-        // Get global node index
-        std::int64_t q = cell_nodes(c, v);
-
-        // Insert (global_vertex_index, local_vertex_index) into map. If
-        // global index seen for first time, add to local-to-global map.
-        auto map_it = global_to_local.insert({q, local_to_global.size()});
-        if (map_it.second)
-          local_to_global.push_back(q);
-
-        // Set local node index in cell node list (applying permutation)
-        cell_nodes_local(c, cell_permutation[v]) = map_it.first->second;
+        // If lowest ranked sharing process is greather than this process,
```

Comment:
I would like to isolate/encapsulate the ownership decision. Making the lowest rank the owner can lead to unbalanced partitions, and will make overlapping computation and communication possibly problematic.

Reply:
Sure. The advantage of using "lowest rank is owner" is that it requires no communication. In any case, ownership is still quite a late decision in the logic of this PR, so this is something which can be easily revisited in the future.